### PR TITLE
Fix fullscreen state re-order by keeping floating clients in the layouts

### LINKED ISSUE #1114: Switching to fullscreen youtube video in chrome moves chrome out of main pain
# Steps
- Using origin/master.
- Using `layout.MonadTall`.
- Open google chrome. Put it in the main pane.
- Play a video on youtube. Press <kbd>f</kbd> to switch to full-screen in the youtube client.
- Press <kbd>esc</kbd> to switch back to "default view" in the youtube client.

# Observed
- google chrome is now a secondary pane.

# Expected
- google chrome should remain in the main pane.

### PR DESCRIPTION
**Pull request by jwijenbergh**
# Overview

This pr is a continuation of #1500 (thanks @whatacold) and aims to fix #1114 #1201 #1202 and #1482. In general with this pr, when a client is set to floating and then tiles again, the position in the layout remains the same. As fullscreening a client essentially involves setting a client to floating, the annoying fullscreen behavior is fixed. I think it's very nice behavior that a client is set to the same position when floating was toggled.

General description of changes for this pr (very similar to #1500)

- ~~`_ClientList` has been renamed to `_TiledClientList` and now has the job of having a list of floating and tiling clients, but only exposes tiled clients to the outside (`tiled_clients` property). `_SimpleLayoutBase` uses this so existing layouts that extend from this do not have to change much or even at all. The name change is there so it more accurately describes what it does imo, but this can be easily reverted.~~ The `_ClientList` name is kept for backwards compatibility. It still consists of tiled and floating clients. To get the tiled clients, use the clients property (same as before).
- When a client is set to floating or tiling, the functions `float_client` and `tile_client` are called respectively. They can be used to correctly set the current client when the current one is set to floating. In the code this is often done by `pseudo` removing/adding clients.
- The layouts themselves have little changes except for the BSP and Tree layouts. The Xmonad layout needs to normalize when a client is set to floating/tiling, this is done by using the `refresh` method from `_SimpleLayoutBase`
- BSP and TreeTab do not use `_SimpleLayoutBase` so they need to manually have a way to skip over floating clients. This code had the least changes as compared to #1500 as I saw no other solution to do it properly.

# FAQ

## Why do layouts have to care about the distinction between floating and tiling clients?

In general they do not. If they use `_SimpleLayoutBase` or even `_ClientList` they can let that do the heavy lifting. However, more involved layouts also have their custom data structure so I see no other way that they must deal with floating clients.

The `tile_client` and `float_client` methods can be used to set a specific client as the active one when the current client is set to floating and back to tiling. This is handled in `_SimpleLayoutBase` but layouts which directly built on top of `_ClientList` need to make sure to call clientlist's `add` and `remove` method with the pseudo flag set to true.

## Why not just remove floating clients and restore the state?

This is an option that is done in #1313. However, while this might work for simple layouts, this does not translate well to more complicated layouts such as BSP and TreeTab. In the end I think this will involve even more and/or complex code then is done here. The way it is in done in this PR is also the way other window managers seem to work, [see e.g. dwm](https://github.com/fanglingsu/dwm/blob/cb3f58ad06993f7ef3a7d8f61468012e2b786cab/dwm.c#L1196)

## Is there a different approach we can take?

There might be. I took some time to think about what the best possible to way to fix this problem is and I just ended up agreeing with the changes made in #1500. While there is actually very little changes that I made myself, I think this is a step in the right direction. I am also very new in the qtile codebase so if anyone has a better idea, I'm all ears and am willing to help. I am, however, quite busy so if someone else wants to continue that is okay too.

# TODO before I set this to be ready for review

- [ ] Tests, both manual and automated. Some automated can be taken from #1500 ?
- [ ] Better descriptions for added/changed methods
- [ ] Fixes for Stack layout 
- [ ] Fixes for scratchpad/dialogs toggling fullscreen
- [x] ~~Are we happy with the code changes for Tree and Bsp? I myself do not use these layouts so also manually testing here is welcome. The Tree code is the only code in here that I do not fully grasp yet~~ I will leave this to the code review
- [x] Odd error when the MonadWide has more than 2 clients and floating and tiling is toggled on and off (`IndexError for relative_sizes list index out of range`).  However, it works as expected. The MonadTall layout does not have this problem for me.
- [x] Make sure layouts that are user written do not have to be changed. This involves calling `add` and `remove` in `tile_client` and `float_client` in the layout code except when the layout is built in
- [x] Fix various issues when windows are shuffled if a window is set to floating/fullscreen

### CODE REVIEW

File: libqtile/group.py [hunk 1]
```diff
@@ -243,31 +245,32 @@ def add(self, win, focus=True, force=False):
...
+                i.remove(win)
 
         if win.floating:
             nextfocus = self.floating_layout.remove(win)
 
+        if win.floating:
             nextfocus = nextfocus or \
                 self.current_window or \
                 self.layout.focus_first() or \
                 self.floating_layout.focus_first(group=self)
```
Comment (m-col) [inline libqtile/group.py:272]: The new line 268 is redundant

**Review by m-col — COMMENTED** [review]
Awesome, and thanks so much for working on this!

> more involved layouts also have their custom data structure so I see no other way that they must deal with floating clients.

To what extent do existing custom layouts (outwith libqtile) need to be updated? Is it possible, or is it already the case, that existing layouts will work but in the current situation where re-tiled windows are appended to the end of their data structure?

Along a similar vein, I think it might be preferable to keep the current name for `_ClientList` and its `clients` attribute, making `clients` what is your `tiled_clients` property. Although yours is definitely more correct, I noticed that one or two of your commits were changing these names in some layouts, and so I think this change may unnecessarily disrupt users who maintain their own layouts. They may be disrupted anyway by other changes, depending on how their layout works, but maybe we can save them that extra bit of hassle.

I haven't tested this yet but will check it out asap. Looks promising! Is there anything you recommend looking at when playing around with it? Just load up all the layouts and toggle a bunch of windows?

**Comment by jwijenbergh** [comment]
> 
> 
> Awesome, and thanks so much for working on this!

No problem!


> To what extent do existing custom layouts (outwith libqtile) need to be updated? Is it possible, or is it already the case, that existing layouts will work but in the current situation where re-tiled windows are appended to the end of their data structure?

It depends on the layout. Provided that they use `_SimpleLayoutBase` they should still work which is most layouts. If someone has its own private layout that that is built from scratch with their own data structure (like Tree or BSP) it would be possible to have the windows appended to the end of their data structure by calling add and remove in `tile_client` and `float_client`, maybe we can make this the default if _SimpleLayoutBase is not used somehow? If they directly use the new qtile with these change they can have problems with their layouts trying to manage floating clients as they're still kept around.

> Along a similar vein, I think it might be preferable to keep the current name for `_ClientList` and its `clients` attribute, making `clients` what is your `tiled_clients` property. Although yours is definitely more correct, I noticed that one or two of your commits were changing these names in some layouts, and so I think this change may unnecessarily disrupt users who maintain their own layouts. They may be disrupted anyway by other changes, depending on how their layout works, but maybe we can save them that extra bit of hassle.

I had that exact thought just now, I will revert those changes.

> I haven't tested this yet but will check it out asap. Looks promising! Is there anything you recommend looking at when playing around with it? Just load up all the layouts and toggle a bunch of windows?

Yes. I recommend just having all the layouts in your config and just play around with it. Toggle/float/fullscreen and see if the expected behavior is there.

File: libqtile/group.py (hunk 1 above)
Comment (jwijenbergh) [reply libqtile/group.py:272]: Thanks! I have seen a few inconsistencies here and there so I will fix them later today 😅

**Comment by jwijenbergh** [comment]
I have force pushed some changes because the history was edited. The _ClientList name is now kept and it honestly makes a lot more sense! The Columns and Stack layouts need no changes now.

File: libqtile/group.py (hunk 1 above)
Comment (jwijenbergh) [reply libqtile/group.py]: Should be fixed now

**Comment by m-col** [comment]
I'm getting some logged errors from monadtall:
```
2021-08-30 09:46:24,608 ERROR libqtile group.py:layout_all():L159 Exception in layout monadtall
Traceback (most recent call last):
  File "/home/mcol/git/qtile/libqtile/group.py", line 157, in layout_all
    self.layout.layout(normal, screen_rect)
  File "/home/mcol/git/qtile/libqtile/layout/base.py", line 50, in layout
    self.configure(i, screen_rect)
  File "/home/mcol/git/qtile/libqtile/layout/xmonad.py", line 319, in configure
    self._configure_specific(client, screen_rect, px, cidx)
  File "/home/mcol/git/qtile/libqtile/layout/xmonad.py", line 357, in _configure_specific
    self.relative_sizes[cidx - 1]
IndexError: list index out of range
```

Also, please could you rebase on top of master?

**Comment by m-col** [comment]
Same error on monadwide too.

**Comment by jwijenbergh** [comment]
> I'm getting some logged errors from monadtall:
> 
> ```
> 2021-08-30 09:46:24,608 ERROR libqtile group.py:layout_all():L159 Exception in layout monadtall
> Traceback (most recent call last):
>   File "/home/mcol/git/qtile/libqtile/group.py", line 157, in layout_all
>     self.layout.layout(normal, screen_rect)
>   File "/home/mcol/git/qtile/libqtile/layout/base.py", line 50, in layout
>     self.configure(i, screen_rect)
>   File "/home/mcol/git/qtile/libqtile/layout/xmonad.py", line 319, in configure
>     self._configure_specific(client, screen_rect, px, cidx)
>   File "/home/mcol/git/qtile/libqtile/layout/xmonad.py", line 357, in _configure_specific
>     self.relative_sizes[cidx - 1]
> IndexError: list index out of range
> ```

Yes this is what I pointed out already, see TODO. That's why I asked if you possible knew what the reason for this is. For me, this does not happen on MonadTall though, how did you reproduce it there?

> Also, please could you rebase on top of master?
Done.

**Comment by m-col** [comment]
> Yes this is what I pointed out already, see TODO. That's why I asked if you possible knew what the reason for this is. For me, this does not happen on MonadTall though, how did you reproduce it there? This will help me narrow down the problem.

Ah yep, apologies. I switched to the layout, created a bunch of windows (it was 5 windows to be exact, in case that influences the outcome), and toggled fullscreen as I moved around the windows. I'll continue testing and see if I can track it down.

**Comment by jwijenbergh** [comment]
> > Yes this is what I pointed out already, see TODO. That's why I asked if you possible knew what the reason for this is. For me, this does not happen on MonadTall though, how did you reproduce it there? This will help me narrow down the problem.
> 
> Ah yep, apologies. I switched to the layout, created a bunch of windows (it was 5 windows to be exact, in case that influences the outcome), and toggled fullscreen as I moved around the windows. I'll continue testing and see if I can track it down.

No problem, just making sure we're on the same page :+1:. I will do the same, it just seems like an odd issue. It's probably an easy fix somewhere though.

**Comment by jwijenbergh** [comment]
I need to rethink some code in this pr to fix some of the open issues, I'm quite busy this week so I'd have to take a look at this mostly next week.

**Comment by m-col** [comment]
Great, no hurry at all.

**Comment by jwijenbergh** [comment]
I had some time to work on this, see the newest commits.

Steps done:

- Fixed lots of shuffling issues by implementing pseudo adding/removing of clients when floating is toggled. This means that the add/remove methods are called but the internal list is not changed. This makes sure so that the internal index/current client in the data structure is set to the same one as when we fully removed the client. This fixes reordering of clients when a client is just set to floating.
- The layout class now by default calls add and remove on tile_client and float_client. This makes sure so that non-builtin layouts need to have 0 changes when this pr is merged. However, these non-builtin layouts will have the previous behaviour of the state not being carried over when a client is set to floating and tiling again. If the user written layouts would want this behaviour then they need to skip over floating clients in their data structure. If they use _SimpleLayoutBase, this is all already handled for them.
- Rebased on top of master
- Fixed Xmonad layout edge case by explicitly checking the relative size bounds. I could not find a better way to fix this issue but everything works as expected here. Whether this was an issue of the code in this pr or some edge case that is triggered by this code, I'm not sure.

This pr is now super useable. I see no issues on my machine as of right now, I will keep testing though. @m-col maybe you still have the same issues?

**Comment by exzombie** [comment]
Trying to float, maximize or fullscreen a window gives me stack traces like these:
```
2021-09-14 08:05:55,956 ERROR libqtile manager.py:process_key_event():L361 KB command error toggle_fullscreen: Traceback (most recent call last):
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/command/interface.py", line 313, in call
    return SUCCESS, cmd(*args, **kwargs)
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1807, in cmd_toggle_fullscreen
    self.toggle_fullscreen()
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1332, in toggle_fullscreen
    self.fullscreen = not self.fullscreen
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1314, in fullscreen
    self._enablefloating(
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1466, in _enablefloating
    self._reconfigure_floating(new_float_state=new_float_state)
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1456, in _reconfigure_floating
    self.group.mark_floating(self, True)
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/group.py", line 294, in mark_floating
    i.float_client(win)
  File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/layout/base.py", line 524, in float_client
    self.refresh()
TypeError: refresh() takes 0 positional arguments but 1 was given
```

**Comment by jwijenbergh** [comment]
> Trying to float, maximize or fullscreen a window gives me stack traces like these:
> 
> ```
> 2021-09-14 08:05:55,956 ERROR libqtile manager.py:process_key_event():L361 KB command error toggle_fullscreen: Traceback (most recent call last):
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/command/interface.py", line 313, in call
>     return SUCCESS, cmd(*args, **kwargs)
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1807, in cmd_toggle_fullscreen
>     self.toggle_fullscreen()
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1332, in toggle_fullscreen
>     self.fullscreen = not self.fullscreen
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1314, in fullscreen
>     self._enablefloating(
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1466, in _enablefloating
>     self._reconfigure_floating(new_float_state=new_float_state)
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/backend/x11/window.py", line 1456, in _reconfigure_floating
>     self.group.mark_floating(self, True)
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/group.py", line 294, in mark_floating
>     i.float_client(win)
>   File "/home/exzombie/.local/lib/python3.8/site-packages/libqtile/layout/base.py", line 524, in float_client
>     self.refresh()
> TypeError: refresh() takes 0 positional arguments but 1 was given
> ```
> 
> I tested Columns and Tiled layouts.

Thanks a lot for testing. Yeah I am stupid sorry, I forgot `self` as the argument for `refresh` :man_facepalming:. Could you re-pull and test? I have also rebased with master.

**Comment by exzombie** [comment]
The problem is gone. Ahh, how sweet this is, I can't wait to see it merged. Thanks a lot for your effort! I'll keep testing and use your branch as a daily driver.

**Comment by jwijenbergh** [comment]
> The problem is gone. Ahh, how sweet this is, I can't wait to see it merged. Thanks a lot for your effort! I'll keep testing and use your branch as a daily driver.

Thanks a lot! If there are any issues just let me know and I will try to fix them asap. I am currently also daily driving it but I don't use all the layouts so some bugs may still be there. However, it should be really doable to use it as a daily driver.

**Comment by VoltaireNoir** [comment]
This is working great so far. Thank you so much for all the effort that you put into this. If I encounter any bugs, I'll surely report them here. Really excited to get back into using Qtile again :D

**Comment by exzombie** [comment]
I observe an issue with the Columns layout that is absent on `master`. I have not seen it with Tile and Matrix layouts.

While maximizing and then unmaximizing a window works very well, closing a maximized window does not. When a maximized window is closed, a hole is left in the layout where the window used to be before it was maximized. Same goes for floating a window or setting it fullscreen. Normalizing does not help, and the hole cannot be easily rid of.

I have not looked at the code, but observing what happens, there is a big difference in behavior between Tile and Columns. When floating a window in Tile, the tiled windows are relaid under the floating window, but the position of the floated window is remembered and it is returned there when unfloated. But with Columns, the tiled windows remain exactly where they are, which is not desired.

**Comment by jwijenbergh** [comment]
> I observe an issue with the Columns layout that is absent on `master`. I have not seen it with Tile and Matrix layouts.
> 
> While maximizing and then unmaximizing a window works very well, closing a maximized window does not. When a maximized window is closed, a hole is left in the layout where the window used to be before it was maximized. Same goes for floating a window or setting it fullscreen. Normalizing does not help, and the hole cannot be easily rid of.
> 
> I have not looked at the code, but observing what happens, there is a big difference in behavior between Tile and Columns. When floating a window in Tile, the tiled windows are relaid under the floating window, but the position of the floated window is remembered and it is returned there when unfloated. But with Columns, the tiled windows remain exactly where they are, which is not desired.

Thanks for reporting. Is the hole the size of a column? Could it be that the column is not removed if only a single client was in that column that is now floating and is removed? If this is the issue you describe then you can move that that column using your layout keybinding to move to the right/left and spawn a new tiling window. I should probably make it so that the column is removed but then that means I would have to remember the column that the window was in somehow.

If this is not the issue, could you give me a snippet of your config that describe the layouts?

**Comment by jwijenbergh** [comment]
@exzombie Please test b7f395d4. I will still have to find a good way to hide columns if it only has floating clients though.

**Comment by exzombie** [comment]
I don't get it. I tested b7f395d, 12e561c, then again c5549a9. I can't reproduce the issue on any of them anymore :( . Sad smiley, because apparently, something more is needed to get qtile into a state where the issue manifests. So I'll keep testing c5549a9 and try to figure it out.

**Comment by jwijenbergh** [comment]
> 
> 
> I don't get it. I tested [b7f395d](https://github.com/qtile/qtile/commit/b7f395d4693b143152bfdcaa320f8f0b3ffc41c9), [12e561c](https://github.com/qtile/qtile/commit/12e561c733212859f7ed47c38360aef6a99555ce), then again [c5549a9](https://github.com/qtile/qtile/commit/c5549a95057a5db056dfb3d9d64a856998feaa36). I can't reproduce the issue on any of them anymore :( . Sad smiley, because apparently, something more is needed to get qtile into a state where the issue manifests. So I'll keep testing [c5549a9](https://github.com/qtile/qtile/commit/c5549a95057a5db056dfb3d9d64a856998feaa36) and try to figure it out.

With c5549a9, try to resize a tiled window in a column with split windows, float it and remove the floating window. Then spawn new windows. You should be seeing errors IIRC.

**Comment by exzombie** [comment]
I have another data point. I was using  c5549a9 until the issue appeared. Then I switched to  b7f395d and restarted qtile. The way the issue manifests is different now: opening a floating dialog window leaves an unselectable empty space, but it disappears when the dialog window is closed. This is better than with c5549a9 where it is very difficult to reclaim the empty space, but is still suboptimal. Opening a floating dialog shouldn't affect the layout.

But I have to say that this is still a bit different than what I observed last week when having a normal window and doing anything "floaty" with it (floating, maximizing, fullscreen) caused the issue. Today, only dialogs work.

I'll keep running b7f395d from now on. BTW, do you think it's acceptable to test stuff by switching to a different commit and restart qtile in-place?

**Comment by jwijenbergh** [comment]
> I have another data point. I was using [c5549a9](https://github.com/qtile/qtile/commit/c5549a95057a5db056dfb3d9d64a856998feaa36) until the issue appeared. Then I switched to [b7f395d](https://github.com/qtile/qtile/commit/b7f395d4693b143152bfdcaa320f8f0b3ffc41c9) and restarted qtile. The way the issue manifests is different now: opening a floating dialog window leaves an unselectable empty space, but it disappears when the dialog window is closed. This is better than with [c5549a9](https://github.com/qtile/qtile/commit/c5549a95057a5db056dfb3d9d64a856998feaa36) where it is very difficult to reclaim the empty space, but is still suboptimal. Opening a floating dialog shouldn't affect the layout.
> 
> But I have to say that this is still a bit different than what I observed last week when having a normal window and doing anything "floaty" with it (floating, maximizing, fullscreen) caused the issue. Today, only dialogs work.
> 
> I'll keep running [b7f395d](https://github.com/qtile/qtile/commit/b7f395d4693b143152bfdcaa320f8f0b3ffc41c9) from now on. BTW, do you think it's acceptable to test stuff by switching to a different commit and restart qtile in-place?

Thanks for the extra details. I cannot reproduce the issue with a dialog window, can you point me in some exact steps to reproduce?

> But I have to say that this is still a bit different than what I observed last week when having a normal window and doing anything "floaty" with it (floating, maximizing, fullscreen) caused the issue. Today, only dialogs work.

Do you mean that only dialogs cause the issue now or only dialogs function correctly?

> I'll keep running b7f395d from now on. BTW, do you think it's acceptable to test stuff by switching to a different commit and restart qtile in-place?

Okay let me know if more issues arise. And it depends on your distro, reproducible systems such as nixos or guix have a specified environment qtile runs in so simply restarting will not update it unless you have qtile patched. For more regular distros like ubuntu/arch restarting should do the trick. 

I recommend testing through Xephyr instead of your main xsession (with a different modifier in the Xephyr session). See https://github.com/qtile/qtile/blob/master/scripts/xephyr for a script.

**Comment by exzombie** [comment]
> Do you mean that only dialogs cause the issue now or only dialogs function correctly?

Sorry, I was unclear. What I meant was that only the dialogs reproduce the problem. However, being back from vacation, I'm now using my work computer again and on this system, both opening a dialog and floating a normal (tiled) window reproduces the problem. I hope to find which difference between the systems is behind this, but I'm not sure yet, there's a lot of variables to check.

In any case, on both machines, I reproduce the problem easily with dialogs:
1. Start a fresh session of b7f395d, use Columns layout (my default).
2. Open firefox, which takes up the whole screen.
3. Press Ctrl-O to get the Open dialog.
4. See the screenshot: firefox is now reduced to the left half of the screen while there is nothing on the right side.
5. When the dialog is closed, firefox is re-tiled to the whole screen. Opening another tiled window and then closing it will also re-tile firefox even while the dialog is open.
image

**Comment by jwijenbergh** [comment]
> > Do you mean that only dialogs cause the issue now or only dialogs function correctly?
> 
> Sorry, I was unclear. What I meant was that only the dialogs reproduce the problem. However, being back from vacation, I'm now using my work computer again and on this system, both opening a dialog and floating a normal (tiled) window reproduces the problem. I hope to find which difference between the systems is behind this, but I'm not sure yet, there's a lot of variables to check.
> 
> In any case, on both machines, I reproduce the problem easily with dialogs:
> 
>     1. Start a fresh session of [b7f395d](https://github.com/qtile/qtile/commit/b7f395d4693b143152bfdcaa320f8f0b3ffc41c9), use Columns layout (my default).
> 
>     2. Open firefox, which takes up the whole screen.
> 
>     3. Press Ctrl-O to get the Open dialog.
> 
>     4. See the screenshot: firefox is now reduced to the left half of the screen while there is nothing on the right side.
> 
>     5. When the dialog is closed, firefox is re-tiled to the whole screen. Opening another tiled window and then closing it will also re-tile firefox even while the dialog is open.
> 
> 
> image

No need to be sorry, I was just making sure. Thanks for the steps, I reproduced it and hopefully fixed it (please test 8349380).

The problem was that I was adding clients that start as floating in the tiling layouts (such as dialogs). This does not make any sense, clients should only be added to the layout if they're tiling, else they need to be added to the `floating_layer` (as is done in master). I have corrected this. 

Sorry for all these different commits and the fact that there are still issues here and there. I do not have a ton of time to test my implementations. If the latest commit works well I will put my efforts into making the git history better (less commits) and add tests so we can actually try to get this in master. I will then also go over the code again fully to see if I can make things simpler or if I see any corner cases left to fix.

**Comment by jwijenbergh** [comment]
There are still lots of issues with the columns layout so I'm not sure if this pr is getting anywhere... Lots of corner cases that need to be handled which is really difficult in the current architecture. I will see if I can fix it but with every fix it seems I just get more issues popping up. If someone else wants to continue it, be my guest :+1:. The columns layout is difficult because we have to remember the structure in a 2d way, skip over columns that have only floating clients and skip over floating clients if a column has tiled clients.

**Comment by jwijenbergh** [comment]
Closing as I think a more simple solution needs to happen
